core/include/vecmem/containers/jagged_device_vector.hpp

```diff
@@ -0,0 +1,171 @@
+         */
+        VECMEM_HOST_AND_DEVICE
+        bool empty(
+            void
+        ) const;
+
+        /**
+         * @brief Get the number of rows in this view.
+         */
+        VECMEM_HOST_AND_DEVICE
+        std::size_t size(
+            void
+        ) const;
+
+        /**
+         * @brief Get the row (vector) at a certain index.
+         *
+         * This method will assert that the element is in range if NDEBUG is
+         * unset.
+         *
+         * @param[in] i The row number to fetch.
+         */
+        VECMEM_HOST_AND_DEVICE
+        device_vector<T> at(
+            std::size_t i
+        );
+
+        /**
+         * @brief Get the row (vector) at a certain index in a const way.
+         *
+         * This method will assert that the element is in range if NDEBUG is
+         * unset.
+         *
+         * @param[in] i The row number to fetch.
+         */
+        VECMEM_HOST_AND_DEVICE
+        const device_vector<T> at(
+            std::size_t i
+        ) const;
+
+        /**
+         * @brief Get the row (vector) at a certain index.
+         *
+         * This method does no bounds checking at all.
+         *
+         * @param[in] i The row number to fetch.
+         */
+        VECMEM_HOST_AND_DEVICE
+        device_vector<T> operator[](
+            std::size_t i
+        );
+
+        /**
+         * @brief Get the row (vector) at a certain index in a const way.
+         *
+         * This method does no bounds checking at all.
+         *
+         * @param[in] i The row number to fetch.
+         */
+        VECMEM_HOST_AND_DEVICE
+        const device_vector<T> operator[](
+            std::size_t i
+        ) const;
+
+        /**
+         * @brief Retrieve the element at a given two-dimensional coordinate.
+         *
+         * This method can be used to directly retrieve an object from an inner
+         * vector.
+         *
+         * @param[in] i The row index to fetch from.
+         * @param[in] j The index in row i to use.
+         */
+        VECMEM_HOST_AND_DEVICE
+        T & at(
+            std::size_t i,
+            std::size_t j
+        );
+
+        /**
+         * @brief Retrieve the element at a given two-dimensional coordinate in
+         * a const way.
+         *
+         * This method can be used to directly retrieve an object from an inner
+         * vector.
+         *
+         * @param[in] i The row index to fetch from.
+         * @param[in] j The index in row i to use.
+         */
+        VECMEM_HOST_AND_DEVICE
+        const T & at(
+            std::size_t i,
+            std::size_t j
+        ) const;
+
+    private:
+        /**
+         * The number of rows in this jagged vector.
+         */
+        const std::size_t m_size;
+
+        /**
+         * The internal state of this jagged vector, which is heap-allocated by
+         * the given memory manager.
+         */
+        details::vector_view<T> * const m_ptr;
```

Comment:
Unlike for `vecmem::details::jagged_vector_view`, here one can indeed make an argument for making these members `const` all the time. But I would still not do it... I would still leave it up to the const-ness of the main object to decide whether these values can be changed or not.

But as I started, on this one I guess I could be persuaded otherwise.

Reply:
I don't like the idea of being able to change these values at all. I have a hard time seeing why it would ever be necessary: if you want to view other data you can simply create another view. Making these non-const simply allows the values to go out of sync more easily for no discernibly upside, as far as I am concerned. :disappointed: